AK/AsyncStream.h

```diff
@@ -0,0 +1,264 @@
+//
+// Calling `peek` when the EOF has been reached as well as calling `read` that tries to read past
+// EOF is a protocol violation and results in EIO error. Next, calling read operations concurrently
+// asserts since it is a logic error. And finally, calling `peek`, `peek_or_eof`, or `read` on a
+// stream that isn't open is a logic error as well and asserts too.
+class AsyncInputStream : public virtual AsyncResource {
+public:
+    struct PeekOrEofResult {
+        ReadonlyBytes data;
+        bool is_eof;
+    };
+
+    AsyncInputStream() = default;
+
+    ReadonlyBytes buffered_data() const
+    {
+        VERIFY(is_open());
+        return buffered_data_unchecked({});
+    }
+
+    Coroutine<ErrorOr<PeekOrEofResult>> peek_or_eof()
+    {
+        VERIFY(is_open());
+        if (!m_is_reading_peek) {
+            m_is_reading_peek = true;
+            auto data = buffered_data_unchecked({});
+            if (!data.is_empty())
+                co_return PeekOrEofResult { data, false };
+        }
+        bool is_not_eof = CO_TRY(co_await enqueue_some({}));
+        co_return PeekOrEofResult { buffered_data_unchecked({}), !is_not_eof };
+    }
+
+    Coroutine<ErrorOr<ReadonlyBytes>> peek()
+    {
+        auto [data, is_eof] = CO_TRY(co_await peek_or_eof());
+        if (is_eof) {
+            reset();
+            co_return Error::from_errno(EIO);
+        }
+        co_return data;
+    }
+
+    Coroutine<ErrorOr<ReadonlyBytes>> read(size_t bytes)
+    {
+        m_is_reading_peek = false;
+
+        auto buffer = buffered_data();
+        while (buffer.size() < bytes) {
+            if (!CO_TRY(co_await enqueue_some({}))) {
+                reset();
+                co_return Error::from_errno(EIO);
+            }
+            buffer = buffered_data_unchecked({});
+        }
+        dequeue({}, bytes);
+        co_return buffer.slice(0, bytes);
+    }
+
+    virtual Coroutine<ErrorOr<bool>> enqueue_some(Badge<AsyncInputStream>) = 0;
+    virtual ReadonlyBytes buffered_data_unchecked(Badge<AsyncInputStream>) const = 0;
+    virtual void dequeue(Badge<AsyncInputStream>, size_t bytes) = 0;
+
+protected:
+    static Badge<AsyncInputStream> badge() { return {}; }
+
+private:
+    bool m_is_reading_peek { false };
+};
+
+class AsyncOutputStream : public virtual AsyncResource {
+public:
+    AsyncOutputStream() = default;
+
+    virtual Coroutine<ErrorOr<size_t>> write_some(ReadonlyBytes buffer) = 0;
+
+    virtual Coroutine<ErrorOr<void>> write(ReadonlySpan<ReadonlyBytes> buffers)
+    {
+        for (auto buffer : buffers) {
+            while (!buffer.is_empty()) {
+                auto nwritten = CO_TRY(co_await write_some(buffer));
+                buffer = buffer.slice(nwritten);
+            }
+        }
+        co_return {};
+    }
+};
+
+class AsyncStream
+    : public AsyncInputStream
+    , public AsyncOutputStream {
+public:
+    AsyncStream() = default;
+};
+
+template<typename T>
```

Comment:
> surely we have concepts for some of these?

I wouldn't expect anything legacy to work with new streams. At the very least, they have a different set of virtual functions to override.

> Or do you want to avoid mixing the headers for Stream/AsyncStream?

That's also true.